test/system/src/test/java/io/pravega/test/system/ConsumptionBasedRetentionWithMultipleReaderGroupsTest.java

```diff
@@ -283,4 +286,140 @@ public void multipleSubscriberCBRTest() throws Exception {
                 5000, 3 * 60 * 1000L);
 
     }
+
+
+    @Test
+    public void updateRetentionPolicyForCBRTest() throws Exception {
+        final ClientConfig clientConfig = Utils.buildClientConfig(controllerURI);
+        assertTrue("Creating scope", streamManager.createScope(SCOPE_1));
+        assertTrue("Creating stream", streamManager.createStream(SCOPE_1, STREAM_1,
+                StreamConfiguration.builder()
+                        .scalingPolicy(ScalingPolicy.fixed(1))
+                        .retentionPolicy(RetentionPolicy.bySizeBytes(MIN_SIZE_IN_STREAM, MAX_SIZE_IN_STREAM)).build()));
+
+        @Cleanup
+        EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(SCOPE_1, clientConfig);
+        @Cleanup
+        EventStreamWriter<String> writer = clientFactory.createEventWriter(STREAM_1, new JavaSerializer<>(),
+                EventWriterConfig.builder().build());
+
+        // Write 7 events to the stream.
+        for (int i = 1; i <= 7; i++) {
+            log.info("Writing event to {}/{}", SCOPE_1, STREAM_1);
+            writer.writeEvent(SIZE_30_EVENT).join();
+        }
+
+        @Cleanup
+        ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(SCOPE_1, clientConfig);
+        ReaderGroupConfig readerGroupConfig = ReaderGroupConfig
+                .builder()
+                .retentionType(ReaderGroupConfig.StreamDataRetention.MANUAL_RELEASE_AT_USER_STREAMCUT)
+                .disableAutomaticCheckpoints().stream(Stream.of(SCOPE_1, STREAM_1)).build();
+
+        assertTrue("Reader group is not created", readerGroupManager.createReaderGroup(READER_GROUP_3, readerGroupConfig));
+        assertEquals(1, controller.listSubscribers(SCOPE_1, STREAM_1).join().size());
+
+        @Cleanup
+        ReaderGroup readerGroup = readerGroupManager.getReaderGroup(READER_GROUP_3);
+        @Cleanup
+        EventStreamReader<String> reader = clientFactory.createReader(READER_GROUP_3 + "-" + 1,
+                READER_GROUP_3, new JavaSerializer<>(), readerConfig);
+
+        EventRead<String> read;
+        // Read three events with reader.
+        for (int i = 1; i <= 3; i++) {
+            read = reader.readNextEvent(READ_TIMEOUT);
+            assertEquals("data of size 30", read.getEvent());
+        }
+        log.info("{} generating 1st stream-cuts for {}/{}", READER_GROUP_3, SCOPE_1, STREAM_1);
+        CompletableFuture<Map<Stream, StreamCut>> futureCuts = readerGroup.generateStreamCuts(streamCutExecutor);
+        // Wait for 5 seconds to force reader group state update. This will allow for the silent
+        // checkpoint event generated as part of generateStreamCuts to be picked and processed.
+        Futures.delayedFuture(Duration.ofSeconds(5), executor).join();
```

Comment:
Can we make this test case independent of 5 sec delay?

Reply:
Giving a delay of 5sec here allows the silent checkpoint generated as part of generateStreamCuts to be picked and processed. The checkpoint will be considered complete and Segments can only be released on the next read call following a checkpoint because this is the only point we can be sure the caller has persisted their position.

Reply:
@AnkurAnand11 having "sleeps" in tests is a bad practice as it builds a fragile testing logic and makes tests unnecessarily long, which lead to longer build times, which leads to lower productivity. The question here is: `futureCuts` is a `CompletableFuture`, would joining that future has a similar result as having to wait for an (arbitrary) period of 5 seconds? If you have done the same in other tests, I would suggest to fix these as well.

Reply:
Here, `readerGroup.generateStreamCuts(streamCutExecutor);` generates a silent checkpoint internally which requires a mandatory followup `reader.readNextEvent(READ_TIMEOUT)` call for the checkpoint to complete and generate a stream cut. If we use `futureCuts.join()` instead, it waits for the `CompletableFuture` to complete and keeps on waiting for the checkpoint to complete which does not happen as we could not call `readNextEvent`. So the delay here is required.

Reply:
I have removed the delay of 5sec and used the logic of artificial clock which is bypassing the wait of 5sec.